### ISSUE TITLE
Enforce proper return types for methods/functions get_order_from_event_body_intent_id

### ISSUE THREAD

**Issue by htdat** [Automattic/woocommerce-payments#4293]
### Description

Based on this convo https://github.com/Automattic/woocommerce-payments/pull/4287#pullrequestreview-984284936

It's more than one time we see fatal errors due to mixed return types.

In this specific case #4287, `get_order_from_event_body_intent_id` returns a few types:

https://github.com/Automattic/woocommerce-payments/blob/caf5ca2298ea77ad4e2b8f8f03c3f037f0a871c6/includes/class-wc-payments-webhook-processing-service.php#L551-L553

But when a new code is added - [ref](https://github.com/Automattic/woocommerce-payments/commit/eb591d24146989485ce9ec65151e3f2e7bb4df95), it misses checking the proper type before using it.

### Acceptance criteria

* Update/refactor `get_order_from_event_body_intent_id` to return less types. Ideally, it would be `null | WC_Order`, and ensure their consumers `process_webhook_payment_intent_failed` and `process_webhook_payment_intent_succeeded` properly check their types before using.
* (Maybe) Check other functions to enforce the same. It's worth checking if PHPCS or Psalm can enforce or prevent this issue. 

### Dev notes

IIRC, Psalm is pretty smart in detecting types and it can prevent this type of issues. Quickly found out this post https://psalm.dev/articles/php-or-type-safety-pick-any-two

### Additional context

This may fit into the RFC improving the plugin code quality paJDYF-3eh-p2

**Comment by htdat** [comment]
cc @vbelolapotkov and @Automattic/transact-architecture as you may want to look at this issue.

**Comment by jbordonado** [comment]
Have you got any thoughts on this before we move this issue to be prioritized? @Automattic/transact-architecture 

Thanks for your input

**Comment by achyuthajoy** [comment]
Checking for update here - p1657623043816659-slack-C0208C3BXHP

**Comment by jessy-p** [comment]
Copied from Slack thread 
```
[12-Jul-2022 04:38:07 UTC] PHP Fatal error:  Uncaught TypeError: Argument 1 passed to WC_Payments_API_Client::deserialize_intention_object_from_array() must be of the type array, string given, called in /var/www/html/wp-content/plugins/woocommerce-payments/includes/wc-payment-api/class-wc-payments-api-client.php on line 471 and defined in /var/www/html/wp-content/plugins/woocommerce-payments/includes/wc-payment-api/class-wc-payments-api-client.php:2342
Stack trace:
#0 /var/www/html/wp-content/plugins/woocommerce-payments/includes/wc-payment-api/class-wc-payments-api-client.php(471): WC_Payments_API_Client->deserialize_intention_object_from_array(' ')
#1 /var/www/html/wp-content/plugins/woocommerce-payments/includes/payment-methods/class-upe-payment-gateway.php(621): WC_Payments_API_Client->get_intent('pi_3LKaorQv0CiK...')
#2 /var/www/html/wp-content/plugins/woocommerce-payments/includes/payment-methods/class-upe-payment-gateway.php(580): WCPay\Payment_Methods\UPE_Payment_Gateway->process_redirect_payment('93', 'pi_3LKaorQv0CiK...', false)
#3 /var/www/ht in /var/www/html/wp-content/plugins/woocommerce-payments/includes/wc-payment-api/class-wc-payments-api-client.php on line 2342
```

**Comment by jessy-p** [comment]
On this issue, let us try to list and resolve all functions with too many return types as given in the example `boolean|WC_Order|WC_Order_Refund`
Moving for prioritisation.

**Comment by RadoslavGeorgiev** [comment]
This issue should only solve the particular instance (`get_order_from_event_body_intent_id` and `boolean|WC_Order|WC_Order_Refund`) and list all other instances of this return type. The new issue should be marked as Priority 3.

**Comment by vbelolapotkov** [comment]
That seems to be part of the payment flow, so routing to your team @deepakpathania for further prioritization.

**Comment by mgascam** [comment]
> This issue should only solve the particular instance (`get_order_from_event_body_intent_id` and `boolean|WC_Order|WC_Order_Refund`) and list all other instances of this return type. The new issue should be marked as Priority 3.

Listing more instances where we have more than two return types in the dockblock in files inside the `includes` folder. I used this [script](https://github.com/mgascam/scan-php-files/blob/main/findUnionTypes.js) to build the list. It scans PHP files with this regex: 

`/@return.*[^\s]+\|[^\s]+\|[^\s]+/`

| File | Line Number |
|------|-------------|
| includes/admin/class-wc-rest-payments-customer-controller.php | 102 |
| includes/admin/class-wc-rest-payments-onboarding-controller.php | 222 |
| includes/admin/class-wc-rest-payments-payment-intents-create-controller.php | 321 |
| includes/admin/class-wc-rest-payments-reader-controller.php | 149 |
| includes/class-experimental-abtest.php | 95 |
| includes/class-wc-payment-gateway-wcpay.php | 656 |
| includes/class-wc-payment-gateway-wcpay.php | 2493 |
| includes/class-wc-payments-db.php | 22 |
| includes/class-wc-payments-db.php | 68 |
| includes/class-wc-payments-db.php | 116 |
| includes/class-wc-payments-payment-request-button-handler.php | 555 |
| includes/class-wc-payments-token-service.php | 65 |
| includes/class-wc-payments-webhook-processing-service.php | 682 |
| includes/class-woopay-tracker.php | 286 |
| includes/express-checkout/class-wc-payments-express-checkout-button-helper.php | 310 |
| includes/express-checkout/class-wc-payments-express-checkout-button-helper.php | 333 |
| includes/reports/class-wc-rest-payments-reports-authorizations-controller.php | 130 |
| includes/reports/class-wc-rest-payments-reports-transactions-controller.php | 118 |
| includes/woopay/class-woopay-session.php | 83 |
| includes/woopay/class-woopay-session.php | 255 |
| includes/woopay/class-woopay-store-api-session-handler.php | 92 |


Please find all the details in the attached zip file (HTML and JSON)

[union-return-types.zip](https://github.com/user-attachments/files/17818491/union-return-types.zip)

**Comment by mgascam** [comment]
I explored using Psalm to prevent issues caused by overly broad union types. As far as I understand, there is no off-the-shelf solution to detect this specific problem. It’s worth noting that a union type, in itself, is not considered an issue by Psalm. Instead, Psalm analyzes how the type is used and highlights any problems only if they are detected. 

I ran Psalm with the error level set to `1` and removed the baseline. Then, I filtered the results to focus on issues related to return types. The most common issue, by far, was `MissingReturnType` (508), followed by `MixedReturnStatement` (187) and `MixedInferredReturnType` (184). Based on these results, I believe we should prioritize addressing the `MissingReturnType` errors first.
